include/expression.hpp

```diff
@@ -0,0 +1,203 @@
+#ifndef ADJACENT_EXPRESSION_HPP
+#define ADJACENT_EXPRESSION_HPP
+
+#include <memory>
+#include <string>
+
+class Expr;
+
+template <class T>
+class Param
+{
+private:
+	T m_value;
+
+public:
+	std::string m_name;
+	bool m_reduceable = true;
+	bool m_changed = false;
+	std::shared_ptr<Expr> m_expr;
+
+	Param() = default;
+	Param(const std::string& name, bool reduceable = true);
+	Param(const std::string& name, double value);
+
+	void set_value(const T& other);
+	T value() const;
+
+	std::shared_ptr<Expr> expr();
+
+	bool operator==(const Param& other) const;
+};
+
+template <class T>
+void Param<T>::set_value(const T& other)
+{
+	if(other == m_value) return;
+	m_changed = true;
+	m_value = other;
+}
+
+template <class T>
+T Param<T>::value() const {
+	return m_value;
+}
+
+template <class T>
+std::shared_ptr<Expr> Param<T>::expr()
+{
+	return m_expr;
+}
+
+template <class T>
+Param<T>::Param(const std::string& name, bool reduceable /* = true */)
+	: m_name(name), m_reduceable(reduceable)
+{
+	m_expr = std::make_shared<Expr>(this);
+}
+
+template <class T>
+Param<T>::Param(const std::string& name, double value)
+	: m_name(name), m_value(value)
+{
+	m_expr = std::make_shared<Expr>(this);
+}
+
+template <class T>
+bool Param<T>::operator==(const Param& other) const
+{
+	return m_name == other.m_name;
+}
+
+namespace std {
+
+  template <>
+  struct hash<Param<double>>
+  {
+    std::size_t operator()(const Param<double>& k) const
+    {
+      // Compute individual hash values for first,
+      // second and third and combine them using XOR
+      // and bit shifting:
+        return std::hash<std::string>()(k.m_name);
```

Comment:
Almost all names will be the same - x, y, z, px, py, pz... This is probably a bad hash function.